pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +254,189 @@ def simulate(
+        return accumulate_native_mcm(circuit, one_shot_meas, [new_shot])
+    for i, m in enumerate(circuit.measurements):
+        if isinstance(m, CountsMP):
+            tmp = Counter(one_shot_meas[i])
+            tmp.update(Counter(new_shot[i]))
+            cat_shot_meas[i] = tmp
+        elif isinstance(m, SampleMP):
+            if not isinstance(one_shot_meas[i], (list, tuple)):
+                cat_shot_meas[i] = [one_shot_meas[i]]
+            else:
+                cat_shot_meas[i] = one_shot_meas[i]
+            cat_shot_meas[i].append(new_shot[i])
+        else:
+            cat_shot_meas[i] = one_shot_meas[i] + new_shot[i]
+    return cat_shot_meas
+
+
+def simulate_native_mcm(
+    circuit: qml.tape.QuantumScript,
+    rng=None,
+    prng_key=None,
+    debugger=None,
+    interface=None,
+) -> Result:
+    """Simulate a single quantum script with native mid-circuit measurements.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. If None, a random key will be
+            generated. Only for simulation using JAX.
+        debugger (_Debugger): The debugger to use
+        interface (str): The machine learning interface to create the initial state with
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+        dict: The mid-circuit measurement results of the simulation
+
+    """
+    state, is_state_batched, mcm_dict = get_final_state(
+        circuit, debugger=debugger, interface=interface
+    )
+    return (
+        measure_final_state(circuit, state, is_state_batched, rng=rng, prng_key=prng_key),
+        mcm_dict,
+    )
+
+
+def has_mid_circuit_measurements(
+    circuit: qml.tape.QuantumScript,
+):
+    """Returns True if the circuit contains a MidMeasureMP object and False otherwise."""
+    return any(isinstance(op, MidMeasureMP) for op in circuit._ops)
+
+
+def requires_mid_samples(measurement):
+    """Determines whether a measurement is a mid-circuit measurement value."""
+    return measurement.mv is not None
+
+
+def idx_sampling_measurements(
+    circuit: qml.tape.QuantumScript,
+):
+    """Returns the indices of sample-like measurements (i.e. CountsMP, SampleMP)."""
+    return [i for i, m in enumerate(circuit.measurements) if requires_mid_samples(m)]
+
+
+def idx_one_shot_measurements(
+    circuit: qml.tape.QuantumScript,
+):
+    """Returns the indices of non sample-like measurements (i.e. not CountsMP, SampleMP)."""
+    return [i for i, m in enumerate(circuit.measurements) if not requires_mid_samples(m)]
+
+
+def gather_native_mid_circuit_measurements(
```

Comment:
Could you rename some of these functions. `gather_native_mid_circuit_measurements`, `gather_statistics`, and `gather_measurements` all have really similar names, as well as really similar doc strings.